src/xenia/base/string_util.h

```diff
@@ -2,224 +2,300 @@
  * Released under the BSD license - see LICENSE in the root for more details. *
  ******************************************************************************
  */
 
 #ifndef XENIA_BASE_STRING_UTIL_H_
 #define XENIA_BASE_STRING_UTIL_H_
 
-#include <cinttypes>
-#include <cstdint>
-#include <cstdlib>
-#include <cstring>
+#include <charconv>
 #include <string>
 
+#include "third_party/fmt/include/fmt/format.h"
+#include "xenia/base/assert.h"
 #include "xenia/base/platform.h"
+#include "xenia/base/string.h"
 #include "xenia/base/vec128.h"
 
+// TODO(gibbed): Clang and GCC don't have std::from_chars for floating point(!)
+// despite it being part of the C++17 standard. Check this in the future to see
+// if it's been resolved.
+
+#if XE_COMPILER_CLANG || XE_COMPILER_GNUC
+#include <cstdlib>
+#endif
+
 namespace xe {
 namespace string_util {
 
-// TODO(gibbed): Figure out why clang doesn't line forward declarations of
-// inline functions.
+inline std::string to_hex_string(uint32_t value) {
+  return fmt::format("{:08X}", value);
+}
 
-std::string to_hex_string(uint32_t value);
-std::string to_hex_string(uint64_t value);
+inline std::string to_hex_string(uint64_t value) {
+  return fmt::format("{:016X}", value);
+}
 
 inline std::string to_hex_string(float value) {
-  union {
-    uint32_t ui;
-    float flt;
-  } v;
-  v.flt = value;
-  return to_hex_string(v.ui);
+  static_assert(sizeof(uint32_t) == sizeof(value));
+  uint32_t pun;
+  std::memcpy(&pun, &value, sizeof(value));
+  return to_hex_string(pun);
 }
 
 inline std::string to_hex_string(double value) {
-  union {
-    uint64_t ui;
-    double dbl;
-  } v;
-  v.dbl = value;
-  return to_hex_string(v.ui);
+  static_assert(sizeof(uint64_t) == sizeof(value));
+  uint64_t pun;
+  std::memcpy(&pun, &value, sizeof(value));
+  return to_hex_string(pun);
 }
 
-std::string to_hex_string(const vec128_t& value);
+inline std::string to_hex_string(const vec128_t& value) {
+  return fmt::format("[{:08X} {:08X} {:08X} {:08X} {:08X}]", value.u32[0],
+                     value.u32[1], value.u32[2], value.u32[3]);
+}
 
-#if XE_ARCH_AMD64
+template <typename T>
+inline T from_string(const std::string_view value, bool force_hex = false) {
+  // Missing implementation for converting type T from string
+  throw;
+}
 
-// TODO(DrChat): This should not exist. Force the caller to use vec128.
-std::string to_hex_string(const __m128& value);
-std::string to_string(const __m128& value);
+namespace internal {
 
-#endif
+template <typename T, typename V = std::make_signed_t<T>>
+inline T make_negative(T value) {
+  if constexpr (std::is_unsigned_v<T>) {
+    value = static_cast<T>(-static_cast<V>(value));
+  } else {
+    value = -value;
+  }
+  return value;
+}
 
+// integral_from_string
 template <typename T>
-inline T from_string(const char* value, bool force_hex = false) {
-  // Missing implementation for converting type T to string
-  throw;
+inline T ifs(const std::string_view value, bool force_hex) {
```

Comment:
Do we have tests for this (and other functions in this header)?

Reply:
Not yet, it's on my TODO. If you'd like to contribute some tests for this PR let me know. 😄